unicon_backend/evaluator/tasks/programming/steps.py

```diff
@@ -573,90 +530,93 @@ def run(
 )
 
 
-class ComputeGraph(Graph[StepClasses]):
-    def _create_link_variable(self, from_node: Step, from_socket: str) -> ProgramVariable:
-        """
-        Create a variable name for the output of a node. The variable name must be unique across all nodes and sockets.
+class ComputeGraph(Graph[StepClasses, GraphEdge[str]]):  # type: ignore
+    VAR_PREFIX: ClassVar[str] = "var_"
+    _var_id: int = PrivateAttr(default=0)
+    _var_id_map: dict[str, int] = PrivateAttr(default={})
+
+    def _get_uniq_var_id(self, node_id: str) -> int:
+        if (var_id := self._var_id_map.get(node_id)) is not None:
+            return var_id
+        self._var_id += 1
+        self._var_id_map[node_id] = self._var_id
+        return self._var_id
+
+    def get_link_var(self, from_node: Step, from_socket: StepSocket) -> ProgramVariable:
+        # Convert into valid Python variable name
+        # Remove all special characters and replace spaces with underscores using regex
+        sanitized_label = re.sub(r"[^a-zA-Z0-9_]", "", from_socket.label.replace(" ", "_"))
+        return cst_var(
+            f"{self.VAR_PREFIX}{self._get_uniq_var_id(from_node.id)}_{from_node.type.value}_{sanitized_label}".upper()
+        )
```

Comment:
does this mean that our socket id is unique is no longer sufficient - and sanitized labels within a node must be unique? we used to check for id uniqueness to ensure the variables were unique

```py
class GraphNode(BaseModel, Generic[IdType, NodeSocketType]):
    id: IdType
    inputs: list[NodeSocketType]
    outputs: list[NodeSocketType]

    @model_validator(mode="after")
    def unique_socket_ids(self) -> Self:
        id_counter = Counter(map(lambda socket: socket.id, chain(self.inputs, self.outputs)))
        if len(id_counter) > 0 and id_counter.most_common(1)[0][1] > 1:
            raise ValueError("Socket ids must be unique")
        return self
```

Reply:
ah my bad this was supposed to be in the final todo

my proposed solution is to add another per socket per node uniq id counter so even if the label name is empty we won't get conflicting var names